commit a8ecdf9cf561a445e61b501186be61a8cd149bf5
Author: Ulf Hermann <ulf_hermann@gmx.net>
Date:   Sat May 25 21:59:18 2013 -0500

    -Fix: return some reserved cargo if vehicle is overloaded by station stopping acceptance for cargo scheduled to be delivered

diff --git a/src/economy.cpp b/src/economy.cpp
index aa7d94c..7ae5dac 100644
--- a/src/economy.cpp
+++ b/src/economy.cpp
@@ -1299,6 +1299,7 @@ static void ReserveConsist(Station *st, Vehicle *u, CargoArray *consist_capleft)
 				continue;
 			}
 
+			assert(v->cargo_cap >= v->cargo.RemainingCount());
 			uint cap = v->cargo_cap - v->cargo.RemainingCount();
 
 			/* Nothing to do if the vehicle is full */
@@ -1398,6 +1399,12 @@ static void LoadUnloadVehicle(Vehicle *front)
 					v->cargo.Reassign(v->cargo.ActionCount(VehicleCargoList::MTA_DELIVER),
 							VehicleCargoList::MTA_DELIVER, VehicleCargoList::MTA_TRANSFER);
 				} else {
+					uint new_remaining = v->cargo.RemainingCount() + v->cargo.ActionCount(VehicleCargoList::MTA_DELIVER);
+					if (v->cargo_cap < new_remaining) {
+						/* Return some of the reserved cargo to not overload the vehicle. */
+						v->cargo.Return(new_remaining - v->cargo_cap, &ge->cargo);
+					}
+
 					/* Keep instead of delivering. This may lead to no cargo being unloaded, so ...*/
 					v->cargo.Reassign(v->cargo.ActionCount(VehicleCargoList::MTA_DELIVER),
 							VehicleCargoList::MTA_DELIVER, VehicleCargoList::MTA_KEEP);
